codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AssembleMiddlewareStack.java

```diff
@@ -41,51 +41,57 @@ public byte getOrder() {
                         .registerMiddleware(MiddlewareRegistrar.builder()
                                 .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
                                         "AddResolveServiceEndpointMiddleware", AwsGoDependency.AWS_MIDDLEWARE)
-                                .build())
+                                        .build())
                                 .useClientOptions()
                                 .build())
                         .build(),
 
-                // Add attemptClockSkew middleware to operation stack
+                // Add unsigned payload middleware to operation stack
                 RuntimeClientPlugin.builder()
+                        .operationPredicate((model, service, operation) -> {
+                            if (service.hasTrait(SigV4Trait.class) && operation.hasTrait(UnsignedPayloadTrait.class)) {
+                                return true;
+                            }
+                            return false;
+                        })
                         .registerMiddleware(MiddlewareRegistrar.builder()
                                 .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
-                                        "AddAttemptClockSkewMiddleware", AwsGoDependency.AWS_MIDDLEWARE)
+                                        "AddUnsignedPayloadMiddleware", AwsGoDependency.AWS_SIGNER_V4)
                                         .build())
                                 .build())
                         .build(),
 
-                // Add newAttempt middleware to operation stack
+                // Add signed payload middleware to operation stack
                 RuntimeClientPlugin.builder()
+                        .operationPredicate((model, service, operation) -> {
+                            if (service.hasTrait(SigV4Trait.class) && !operation.hasTrait(UnsignedPayloadTrait.class)) {
+                                return true;
+                            }
+                            return false;
+                        })
                         .registerMiddleware(MiddlewareRegistrar.builder()
-                                .resolvedFunction( SymbolUtils.createValueSymbolBuilder(
-                                        "AddRetryMiddlewares", AwsGoDependency.AWS_RETRY)
+                                .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
+                                        "AddComputePayloadSHA256Middleware", AwsGoDependency.AWS_SIGNER_V4)
```

Comment:
Should AddComputePayloadSHA256Middleware be before UnsignedPayload Middleware addition?

Reply:
Ideally they are mutually exclusive, so cannot be present at the same time.

Reply:
Oh I thought the change was required as they can exist together. Why separate it from HTTPSignerMiddlewares then?

Reply:
Mainly because only one could be used at a time. Either Checksum or unsigned middleware, if both were present the wrong behavior might occur.  This was the reasoning for breaking out checksum middleware from the signer middleware. So that the signer could always be used separately from how the payload checksum is or is not set.